config_explained.md

# Newt's __newt config__ Explained

The __config__ action can create a new skeleton Newt YAML file. It can be used to update an applications configuration and applications metadata. It allows you to select which applications are known by your data pipeline as well as set basic runtime configuration which is used by the `newt run` command. In addition you can use Newt Config to update the configurations in the applications' property. You might need to do this if the standard ports for the network services need to be changed.

## Create your first Newt YAML file

By default if you do not provide a YAML filename the `newt config` command will use the filename of "app.yaml". Let's try creating a "app.yaml" by running the config action.

~~~shell
newt config app.yaml
~~~

You will now get prompted by a series of questions that will configure what is listed inside the applications' property.

Here' is an example of me running the command you can play through.

FIXME: Insert Asciienema demo of run `newt config app.yaml` here.

FIXME: Explain the menu system and inactive flow of question.


~~~yaml
#!/usr/bin/env newt check
#
# This was generated by rsdoiel on 2024-08-08 with newt version 0.0.8 ea131b9.
#
app_metadata:
  name: app
  title: This is a tutorial app using Newt
  copyright: 2024
  logo_link: /media/logo.svg
  logo_text: Logo of the demo app
  header_text: A Newt Turtorial YAML file
  header_link: https://caltechlibrary.github.io/newt
  contact_address: 123 Calle Quattro, Colonia Major, Estat de Providencia, Republic of the Collective Imagination, 456789
applications:
  router:
    port: 8010
  template_engine:
    port: 8011
    base_dir: views
    partials_dir: partials
    ext_name: .hbs
  postgres:
    namespace: app
    port: 5432
    dsn: postgres://{PGUSER}:{PGPASSWORD}@localhost:5432/app.yaml
  postgrest:
    app_path: postgrest
    conf_path: postgrest.conf
    port: 3000
  enviroment:
    - PGUSER
    - PGPASSWORD
~~~

When I accepted using Postgres the config action knows that the connection string can be built from the standard environment variables expected by Postgres' psql and database connectors.  That is why even through I answered "N" to the question about setting up environment variables to import I still have environment variables listed.

NOTE: The config option can be used to update an existing Newt YAML file, e.g. "app.yaml". If you change the configuration then you'll be prompted to save the changes as the last question. If you save the changes the prior file will be saved to a ".bak" version before writing the changes to disk. If you want to abort your changes completely then you can press "control C" to quit at any time before the final save prompt.

Next up [Newt Model](model_explained.md)


newt_config_explained.md

# Newt's `newt config`

The Newt config action can create a new skeleton YAML file or update the application's property of a Newt YAML. It allows you to select which application is run by `newt run` as well as configure those applications. In addition you can use Newt Config to update the configurations in the applications' property. You might need to do this if the standard ports for the network services need to be changed.

## Create your first Newt YAML file

By default if you do not provide a YAML filename the `newt config` command will use the filename of "app.yaml". Let's try creating a "app.yaml" by running the config action.

~~~shell
newt config
~~~

You will now get prompted by a series of questions that will configure what is listed inside the applications' property.

This is a transcript of running `newt config` to generated "app.yaml".

~~~shel
newt config
Creating "app.yaml"
Will app.yaml use Newt Router (Y/n)? Y
Will app.yaml use Newt Router (Y/n)? Y
Will app.yaml use Postgres (Y/n)? Y
Will app.yaml use PostgREST (Y/n)? Y
Will app.yaml use Newt Mustache (Y/n)? Y
Will app.yaml use Newt Generator (Y/n)? Y
Will app.yaml need to import environment variables (y/N)? N
Will app.yaml provide options to the services (y/N)? N
Save and exit (Y/n)? Y
~~~

The single letter, "Y" or "N", at the end of each question is the answer I provided when generating the following "app.yaml". If you just press enter rather then specifically answer "Y" or "N" then you are accepting the defaults of generating the application configuration. This is fine in most cases.

Here's the generated "app.yaml".

~~~yaml
#/usr/bin/env newt check
#
# This was generated by rsdoiel on 2024-05-09 with newt version 0.0.8 bc0d5e8.
#
applications:
  newtrouter:
    port: 8010
  newtmustache:
    port: 8011
  newtgenerator:
    namespace: app.yaml
  postgres:
    port: 5432
    dsn: postgres://{PGUSER}:{PGPASSWORD}@localhost:5432/app.yaml
  postgrest:
    app_path: postgrest
    conf_path: postgrest.conf
    port: 3000
  enviroment:
    - PGUSER
    - PGPASSWORD
~~~

When I accepted using Postgres the config action knows that the connection string can be built from the standard environment variables expected by Postgres' psql and database connectors.  That is why even through I answered "N" to the question about setting up environment variables to import I still have environment variables listed.

NOTE: if "app.yaml" already existed then answer "Y" or "N" questions would have only updated "app.yaml" after I answered "Y" to the question, "Save and exit (Y/n)?". Before that any changes would have been in the computer's memory but not acted upon by writing them to disk.

NOTE: When you are running `newt config` you can press "control C" to quit the "config" action without writing the YAML to disk.

Next up [Newt Model](newt_model_explained.md)



cli.go
/**
 * cli.go an implements runners for the cli of the Newt Project.
 *
 * @author R. S. Doiel
 */
package newt

import (
	"bufio"
	//	"bytes"
	//	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"os/signal"
	"path"
	"path/filepath"
	"strings"
	"time"
)

const (
	// These constants are used for exit code. FIXME: look up the POSIX recommendation on exit
	// codes and adopt those.
	OK = iota
	CONFIG

	// General failure of a command or service
	INIT_FAIL
	CHECK_FAIL
	MODELER_FAIL
	GENERATOR_FAIL
	ROUTER_FAIL
	TEMPLATE_ENGINE_FAIL
	NEWT_FAIL
	SWS_FAIL
	POSTGREST_FAIL

	// Internal service failures
	RESOLVE
	HANDLER
	SERVER_ERROR
	UNSUPPORTED_ACTION
	DATA_ERROR
	READ_ERROR
	DECODE_ERROR
	TEMPLATE_ERROR

	// Default service settings
	ROUTER_PORT             = 8010
	TEMPLATE_ENGINE_PORT    = 8011
	TEMPLATE_ENGINE_TIMEOUT = 3 * time.Second
	TEMPLATE_ENGINE_BASE_DIR   = "views"
	TEMPLATE_ENGINE_PARTIALS_DIR   = "partials"
	TEMPLATE_ENGINE_EXT_NAME   = ".hbs"
	SWS_PORT                = 8000
	SWS_HTDOCS              = "."
	POSTGREST_PORT          = 3000
	POSTGRES_PORT           = 5432
)

// backupFile takes a filename, copies it to filename plus ".bak"
func backupFile(appFName string) error {
	buf, err := os.ReadFile(appFName)
	if err != nil {
		return fmt.Errorf("failed to back up %s, aborting write, %s", appFName, err)
	}
	if err := os.WriteFile(appFName+".bak", buf, 0666); err != nil {
		return fmt.Errorf("failed to write back up %s, aborting write, %s\n", appFName, err)
	}
	return nil
}

// getNewtYamlFName - figure out what the Newt YAML filename should be.
// If no filename is provided use the default "app.yaml".
func getNewtYamlFName(args []string) string {
	fName := ""
	for _, arg := range args {
		arg = strings.TrimSpace(arg)
		if arg != "" && !strings.HasPrefix(arg, "-") {
			fName = arg
			break
		}
	}
	if fName == "" {
		fName = "app.yaml"
	}
	return fName
}

// hasArg - review args and see if the use option is in the list. If
// hasArg is found then true is returned if not false.
func hasArg(option string, args []string) bool {
	for _, arg := range args {
		if strings.ToLower(arg) == option {
			return true
		}
	}
	return false
}

func renderTemplate(generator *Generator, tType string, modelID string, action string, fName string) error {
	var err error
	dName := path.Dir(fName)
	if dName != "" && dName != "." {
		if _, err = os.Stat(fName); os.IsNotExist(err) {
			os.MkdirAll(dName, 0775)
		}
	}
	if _, err = os.Stat(fName); err == nil {
		if err = backupFile(fName); err != nil {
			return err
		}
	}
	out, err := os.Create(fName)
	if err != nil {
		return err
	}
	defer out.Close()
	generator.out = out
	if err := generator.Generate(tType, modelID, action); err != nil {
		return err
	}
	return nil
}

// RunGenerator is a runner for generating SQL and templates from our Newt YAML file.
func RunGenerator(in io.Reader, out io.Writer, eout io.Writer, args []string) int {
	fName := getNewtYamlFName(args)
	if fName == "" {
		fmt.Fprintf(eout, "missing Newt YAML filename\n")
		return CONFIG
	}

	ast, err := LoadAST(fName)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return CONFIG
	}
	generator, err := NewGenerator(ast)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return GENERATOR_FAIL
	}
	//NOTE: I need to generate each of the files needed for Postgres and PostgREST
	for _, fName := range []string{"setup.sql", "models.sql"} {
	    action := strings.TrimSuffix(fName, ".sql")	
		if err := renderTemplate(generator, "postgres", "", action, fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
	}

	fName = "postgrest.conf"
	if err := renderTemplate(generator, "postgrest", "", "", fName); err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return GENERATOR_FAIL
	}

	//NOTE: For each model generate a set of templates
	for _, modelID := range ast.GetModelIds() {
		// backup and generate {model}_create_form.hbs, {model}_create_response.hbs
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_create_form%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "create_form", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_create_response%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "create_response", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}

		// backup and generate {model}_read.hbs
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_read%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "read", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
		// backup and generate {model}_update_form.hbs, {model}_update_response.hbs
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_update_form%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "update_form", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_update_response%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "update_response", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
		// backup and generate {model}_delete_form.hbs, {model}_delete_response.hbs
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_delete_form%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "delete_form", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_delete_response%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "delete_response", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
		// backup and generate {model}_list.hbs
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, fmt.Sprintf("%s_list%s", modelID, ast.Applications.TemplateEngine.ExtName))
		if err := renderTemplate(generator, "template", modelID, "list", fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
	}
	// Need to generate Handlebars partials.
	for _, partialType := range []string{ "head", "header", "nav", "footer" } {
		fName = path.Join(ast.Applications.TemplateEngine.BaseDir, ast.Applications.TemplateEngine.PartialsDir, partialType + ast.Applications.TemplateEngine.ExtName)
		if err := renderTemplate(generator, "partial_template", "", partialType, fName); err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return GENERATOR_FAIL
		}
	}
	return OK
}

// RunTemplateEngine is a runner for a Newt's template engine.
func RunTemplateEngine(in io.Reader, out io.Writer, eout io.Writer, args []string, port int, timeout int, verbose bool) int {
	appName := "Newt Template Engine"
	fName := getNewtYamlFName(args)
	if fName == "" {
		fmt.Fprintf(eout, "missing Newt YAML filename\n")
		return CONFIG
	}
	// Load the Newt YAML syntax file holding the configuration
	// and make sure it conforms.
	ast, err := LoadAST(fName)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return CONFIG
	}
	if ast.Applications == nil {
		fmt.Fprintf(eout, "%s not configed in %s\n", appName, fName)
		return CONFIG
	}
	if ast.Templates == nil {
		fmt.Fprintf(eout, "now templates found in %s\n", fName)
		return CONFIG
	}
	if ast.Applications.TemplateEngine == nil {
		fmt.Fprintf(eout, "missing template engine configuration in %q", fName)
		return CONFIG
	}
	// Instantiate the specific application with the filename and AST object
	mt, err := NewTemplateEngine(ast)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return CONFIG
	}

	// If port is not set in the config, set it to the default port.
	if mt.Port == 0 {
		mt.Port = TEMPLATE_ENGINE_PORT
	}
	// See if we have a command line option for port to process.
	if port != 0 {
		mt.Port = port
	}
	if verbose {
		fmt.Fprintf(out, "port set to %d\n", mt.Port)
	}

	if timeout != 0 {
		mt.Timeout = time.Duration(timeout) * time.Second
	}
	if mt.Timeout == 0 {
		mt.Timeout = TEMPLATE_ENGINE_TIMEOUT
	}
	if len(mt.Templates) == 0 {
		fmt.Fprintf(eout, "no templates in configuration\n")
		return CONFIG
	}
	// Create mux for http service
	// Resolve partial templates and build handlers
	for _, tmpl := range mt.Templates {
		if verbose {
			tmpl.Debug = true
		}
		if err := tmpl.ResolveTemplate(); err != nil {
			fmt.Fprintf(eout, "%s failed to resolve, %s\n", tmpl.Template, err)
			return RESOLVE
		}
		if err := tmpl.ResolvePath(); err != nil {
			fmt.Fprintf(eout, "failed to build handler for %q, %s\n", tmpl.Pattern, err)
			return HANDLER
		}
	}
	// Launch web service
	fmt.Printf("starting %s listening on port :%d (press Ctrl-c to exit)\n", appName, mt.Port)
	if err := mt.ListenAndServe(); err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return TEMPLATE_ENGINE_FAIL
	}
	return OK
}

// RunRouter is a runner for Newt data router and static file service
func RunRouter(in io.Reader, out io.Writer, eout io.Writer, args []string, dryRun bool, port int, htdocs string, verbose bool) int {
	appName := "Newt Router"
	// You can run Newt Router with just an htdocs directory. If so you don't require a config file.
	var (
		err    error
		router *Router
	)
	fName := getNewtYamlFName(args)
	if fName == "" {
		fmt.Fprintln(eout, "missing Newt YAML filename")
		return CONFIG
	}
	ast, err := LoadAST(fName)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return CONFIG
	}
	// Finally Instantiate the router from fName and AST object
	router, err = NewRouter(ast)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return CONFIG
	}
	if router.Port == 0 {
		router.Port = ROUTER_PORT
	}
	if port != 0 {
		router.Port = port
	}
	if htdocs != "" {
		router.Htdocs = htdocs
	}

	// Are we ready to run service?
	if router.Routes == nil && router.Htdocs == "" {
		fmt.Fprintf(eout, "nether routes or htdocs are set.")
		return CONFIG
	}

	if router.Port == 0 {
		fmt.Fprintf(eout, "port is not set, default is not available\n")
		return CONFIG
	}

	if verbose && router.Routes != nil {
		for _, route := range router.Routes {
			route.Debug = true
		}
	}

	if dryRun {
		fmt.Fprintf(out, "configuration and routes successfully processed\n")
		return OK
	}

	// Launch web services
	fmt.Fprintf(out, "starting %s listening on port :%d (use Ctr-c to exit)\n", appName, router.Port)
	if router.Htdocs != "" {
		dir, err := filepath.Abs(router.Htdocs)
		if err == nil {
			fmt.Fprintf(out, "\tstatic content %s\n", dir)
		} else {
			fmt.Fprintf(out, "\tstatic content %s (warning: %s)\n", router.Htdocs, err)
		}
	}
	if err := router.ListenAndServe(); err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return ROUTER_FAIL
	}
	return OK
}

// RunStaticWebServer this provides a localhost for static file content.
func RunStaticWebServer(in io.Reader, out io.Writer, eout io.Writer, args []string, port int, verbose bool) int {
	appName := "Newt Static Web Server"
	if port == 0 {
		port = SWS_PORT
	}
	htdocs := SWS_HTDOCS
	if len(args) > 0 {
		htdocs = args[0]
	}
	fmt.Fprintf(out, "starting %s listening on port :%d (use Ctr-c to exit)\n", appName, port)
	if err := NewtStaticFileServer(port, htdocs, verbose); err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return SWS_FAIL
	}
	return OK
}

// NewtRunCheckYAML will load a Newt YAML fiel and make sure it can parse the configuration.
func RunNewtCheckYAML(in io.Reader, out io.Writer, eout io.Writer, args []string, verbose bool) int {
	fName := getNewtYamlFName(args)
	if fName == "" {
		fmt.Fprintln(eout, "missing Newt YAML filename")
		return CHECK_FAIL
	}
	ast, err := LoadAST(fName)
	if err != nil {
		fmt.Fprintf(eout, "%s error: %s\n", fName, err)
		return CHECK_FAIL
	}
	if !ast.Check(eout) {
		return CHECK_FAIL
	}
	if verbose {
		if ast.Applications.PostgREST != nil {
			fmt.Fprintf(out, "PostgREST configuration is %s\n", ast.Applications.PostgREST.ConfPath)
			fmt.Fprintf(out, "PostgREST will be run with the command %q\n", strings.Join([]string{
				ast.Applications.PostgREST.AppPath,
				ast.Applications.PostgREST.ConfPath,
			}, " "))
		}
		if ast.Models != nil {
			for _, m := range ast.Models {
				fmt.Printf("models %s defined, %d elements\n", m.Id, len(m.Elements))
				if m.Description != "" {
					fmt.Fprintf(out, "\t%s\n\n", m.Description)
				}
			}
		}
		if ast.Applications.Router != nil {
			port := ROUTER_PORT
			if ast.Applications.Router.Port != 0 {
				port = ast.Applications.Router.Port
			}
			fmt.Fprintf(out, "Newt Router configured, port set to :%d\n", port)
			if ast.Applications.Router.Htdocs != "" {
				fmt.Fprintf(out, "Static content will be served from %s\n", ast.Applications.Router.Htdocs)
			}
			if ast.Routes != nil {
				for _, r := range ast.Routes {
					fmt.Fprintf(out, "route %s defined, request path %s, pipeline size %d\n", r.Id, r.Pattern, len(r.Pipeline))
					if r.Description != "" {
						fmt.Fprintf(out, "\t%s\n\n", r.Description)
					}
				}
			}
		}
		if ast.Applications.TemplateEngine != nil {
			port := ast.Applications.TemplateEngine.Port
			if port == 0 {
				port = TEMPLATE_ENGINE_PORT
			}
			fmt.Fprintf(out, "Newt template engine configured, port set to :%d\n", port)
			fmt.Fprintf(out, "%d templates are defined\n", len(ast.Templates))
			for _, mt := range ast.Templates {
				tList := []string{
					mt.Template,
				}
				fmt.Fprintf(out, "http://localhost:%d%s points at %s\n", port, mt.Pattern, strings.Join(tList, ", "))
				if mt.Description != "" {
					fmt.Fprintf(out, "\t%s\n\n", mt.Description)
				}
			}
		}
	}
	return OK
}

// RunNewt is a runner that can run Newt's router and template engine plus PostgREST if defined in the Newt YAML file.
func RunNewt(in io.Reader, out io.Writer, eout io.Writer, args []string, verbose bool) int {
	appName := path.Base(os.Args[0])
	action := ""
	// Extract our actions from the args
	switch len(args) {
	case 0:
		return NEWT_FAIL
	case 1:
		action, args = args[0], []string{}
	default:
		action, args = args[0], args[1:]
	}

	switch action {
	case "config":
		return RunNewtConfig(in, out, eout, args, verbose)
	case "check":
		return RunNewtCheckYAML(in, out, eout, args, verbose)
	case "model":
		return RunModeler(in, out, eout, args)
	case "generate":
		return RunGenerator(in, out, eout, args)
	case "run":
		return RunNewtApplications(in, out, eout, args, verbose)
	case "ws":
		return RunStaticWebServer(in, out, eout, args, 0, verbose)
	default:
		fmt.Fprintf(eout, "%s does %q is an unsupported action, see %s -help\n", appName, action, appName)
		return UNSUPPORTED_ACTION
	}
	return OK
}

// RunNewtApplications will run the applictions defined in your Newt YAML file.
func RunNewtApplications(in io.Reader, out io.Writer, eout io.Writer, args []string, verbose bool) int {
	appName := path.Base(os.Args[0])
	// Get the Newt YAML file to run
	fName := getNewtYamlFName(args)
	if fName == "" {
		fmt.Fprintf(eout, "%s expected a Newt YAML filename\n", appName)
		return CONFIG
	}
	ast, err := LoadAST(fName)
	if err != nil {
		fmt.Fprintf(eout, "%s failed to load %q, %s", appName, fName, err)
		return CONFIG
	}
	if ast.Applications == nil {
		fmt.Fprintf(eout, "no applications configured in %s", fName)
		return CONFIG
	}
	// Startup PostgREST if configured in the Newt YAML file.
	if ast.Applications.PostgREST != nil &&
		ast.Applications.PostgREST.ConfPath != "" && ast.Applications.PostgREST.AppPath != "" {
		postgREST := ast.Applications.PostgREST
		cwd, err := os.Getwd()
		if err != nil {
			log.Println(err)
			return POSTGREST_FAIL
		}
		cmd := exec.Command(postgREST.AppPath, postgREST.ConfPath)
		cmd.Dir = cwd
		// Setup the stdin and stdout to be visible from Newt
		cmd.Stdout = out
		cmd.Stderr = eout
		log.Printf("starting %s %s listening on :%d (use Ctrl-c to exit)", postgREST.AppPath, postgREST.ConfPath, postgREST.Port)
		err = cmd.Start()
		if err != nil {
			log.Println(err)
			return POSTGREST_FAIL
		}
		log.Printf("%s running with pid %d in the backround", postgREST.AppPath, cmd.Process.Pid)
		cmd.Process.Release()
	}
	// Setup and start Newt template engine first
	if ast.Applications.TemplateEngine != nil {
		go func() {
			RunTemplateEngine(in, out, eout, args, 0, 0, verbose)
		}()
	}

	// The router starts up second and is what prevents service from falling through.
	if ast.Applications.Router != nil {
		go func() {
			RunRouter(in, out, eout, args, false, 0, "", verbose)
		}()
	}
	// NOTE: we need to wait for a signal so that our external process and Go routines aan run.

	// Set up channel on which to send signal notifications.
	c := make(chan os.Signal, 1)

	// We're listening for all signals, probably should narrow this down.
	signal.Notify(c, os.Interrupt, os.Kill)

	// Block until any signal is received.
	s := <-c
	fmt.Println("exited with signal:", s)
	return OK
}

// RunNewtConfig will initialize a Newt project by creating a Newt YAML file interactively.
func RunNewtConfig(in io.Reader, out io.Writer, eout io.Writer, args []string, verbose bool) int {
	var answer string

	ast := &AST{}
	readBuffer := bufio.NewReader(in)
	// Step 1. Figure out what we're going to call our generated Newt YAML file.
	appFName := getNewtYamlFName(args)
	if appFName == "" {
		fmt.Fprintf(eout, "missing Newt YAML Filename\n")
		return INIT_FAIL
	}
	skipPrompts := hasArg("-y", args)
	if skipPrompts {
		answer = "y"
	}
	if !skipPrompts {
		if _, err := os.Stat(appFName); err == nil {
			fmt.Fprintf(out, "Opening %q\n", appFName)
			ast, err = LoadAST(appFName)
		} else if len(args) <= 1 {
			fmt.Fprintf(out, "Creating %q\n", appFName)
		}
	}
	// Step 2. Decide about meta data of your app
	if ast.AppMetadata == nil {
		ast.AppMetadata = new(AppMetadata)
	}
	// Step 3. Decide which services you're going to use (a .Applications will need to exist).
	if ast.Applications == nil {
		ast.Applications = new(Applications)
	}
	for {
		//NOTE: Each of these should reflect the current model list in ast.
		setupAppMetadata(ast, readBuffer, out, appFName, skipPrompts)
		setupRouter(ast, readBuffer, out, appFName, skipPrompts)
		setupPostgresAndPostgREST(ast, readBuffer, out, appFName, skipPrompts)
		setupTemplateEngine(ast, readBuffer, out, appFName, skipPrompts)
		setupEnvironment(ast, readBuffer, out, appFName, skipPrompts)
		setupOptions(ast, readBuffer, out, appFName, skipPrompts)

		// Now output the YAML
		_, err := ast.Encode()
		if err != nil {
			fmt.Fprintf(eout, "Failed to generate %s, %s\n", appFName, err)
			return INIT_FAIL
		}
		if skipPrompts {
			answer = "y"
		} else {
			fmt.Fprintf(out, "Save and exit (Y/n)? ")
			answer = getAnswer(readBuffer, "y", true)
		}
		if answer == "y" {
			// We're ready to write out result.
			// If file exists make a back up copy
			if err := ast.SaveAs(appFName); err != nil {
				fmt.Fprintf(eout, "failed to write %s, %s\n", appFName, err)
				return INIT_FAIL
			}
			break
		} else {
			fmt.Fprintf(out, "Exit without saving (y/N)? ")
			answer = getAnswer(readBuffer, "n", true)
			if answer == "y" {
				fmt.Fprintf(out, "%s was not saved\n", appFName)
				return INIT_FAIL
			}
		}
	}
	if _, err := os.Stat(".git"); err == nil {
		fmt.Fprintf(out, `It appears your are using the git revision control system.
You should make sure that generated code containing secrets is NOT included in your
repository for %q. It is recommented that add the following to your .gitignore file.

    # Newt Project ignore list.
    *setup*.sql
    postgrest.conf

`, appFName)
	}
	return OK
}

func RunModeler(in io.Reader, out io.Writer, eout io.Writer, args []string) int {
	var (
		answer string
	)
	ast := &AST{}
	readBuffer := bufio.NewReader(in)
	// Step 1. Figure out what we're going to call our generated Newt YAML file.
	appFName := getNewtYamlFName(args)
	if appFName == "" {
		fmt.Fprintf(eout, "missing Newt YAML Filename\n")
		return MODELER_FAIL
	}
	if _, err := os.Stat(appFName); err == nil {
		ast, err = LoadAST(appFName)
		if err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			return MODELER_FAIL
		}
	} else {
		fmt.Fprintf(out, "Create %q (Y/n)? ", appFName)
		answer = getAnswer(readBuffer, "y", true)
		if answer != "y" {
			fmt.Fprintf(eout, "aborting creation of %q\n", appFName)
			return MODELER_FAIL
		}
	}
	// Step 2. build our lists of models and manage them
	if err := modelerTUI(ast, in, out, eout, appFName, args); err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		return MODELER_FAIL
	}
	if ast.Applications == nil ||
		ast.Applications.Router == nil ||
		ast.Applications.Postgres == nil ||
		ast.Applications.PostgREST == nil ||
		ast.Applications.TemplateEngine == nil {
		fmt.Fprintf(out, "Applications are not configured for %q, try\n\n", appFName)
		appName := path.Base(os.Args[0])
		fmt.Fprintf(out, "\t%s config %q\n\n", appName, appFName)
	}
	return OK
}
